File: relmgmt/frontend/src/services/api/v1/componentService.ts
import { apiClient } from '../apiClient';
import { 
  Component, 
  ComponentRequest, 
  ComponentResponse, 
  ComponentTypeEnum,
  ComponentType,
  PaginatedResponse 
} from '../../../types';

/**
 * Service for managing components within scope items
 */
export class ComponentService {
  private static readonly BASE_URL = '/components';
  private static readonly SCOPE_ITEMS_URL = '/scope-items';

  /**
   * Get all components for a scope item
   */
  static async getComponentsByScopeItemId(scopeItemId: number): Promise<ComponentResponse[]> {
    const response = await apiClient.get<ComponentResponse[]>(
      `${this.SCOPE_ITEMS_URL}/${scopeItemId}/components`
    );
    return response.data;
  }

  /**
   * Get a component by ID
   */
  static async getComponent(id: number): Promise<ComponentResponse> {
    const response = await apiClient.get<ComponentResponse>(`${this.BASE_URL}/${id}`);
    return response.data;
  }

  /**
   * Create a new component for a scope item
   */
  static async createComponent(scopeItemId: number, component: ComponentRequest): Promise<ComponentResponse> {
    const response = await apiClient.post<ComponentResponse>(
      `${this.SCOPE_ITEMS_URL}/${scopeItemId}/components`,
      component
    );
    return response.data;
  }

  /**
   * Update an existing component
   */
  static async updateComponent(id: number, component: ComponentRequest): Promise<ComponentResponse> {
    const response = await apiClient.put<ComponentResponse>(`${this.BASE_URL}/${id}`, component);
    return response.data;
  }

  /**
   * Delete a component
   */
  static async deleteComponent(id: number): Promise<void> {
    await apiClient.delete(`${this.BASE_URL}/${id}`);
  }

  /**
   * Get components by component type
   */
  static async getComponentsByType(componentType: ComponentTypeEnum): Promise<ComponentResponse[]> {
    const response = await apiClient.get<ComponentResponse[]>(
      `${this.BASE_URL}/type/${componentType}`
    );
    return response.data;
  }

  /**
   * Get components by scope item ID and component type
   */
  static async getComponentsByScopeItemAndType(
    scopeItemId: number, 
    componentType: ComponentTypeEnum
  ): Promise<ComponentResponse[]> {
    const response = await apiClient.get<ComponentResponse[]>(
      `${this.SCOPE_ITEMS_URL}/${scopeItemId}/components/type/${componentType}`
    );
    return response.data;
  }

  /**
   * Count components by scope item ID
   */
  static async countComponentsByScopeItemId(scopeItemId: number): Promise<number> {
    const response = await apiClient.get<number>(
      `${this.SCOPE_ITEMS_URL}/${scopeItemId}/components/count`
    );
    return response.data;
  }

  /**
   * Check if a component can be deleted
   */
  static async canDeleteComponent(id: number): Promise<boolean> {
    const response = await apiClient.get<boolean>(`${this.BASE_URL}/${id}/can-delete`);
    return response.data;
  }

  /**
   * Get components by release ID
   */
  static async getComponentsByReleaseId(releaseId: number): Promise<ComponentResponse[]> {
    const response = await apiClient.get<ComponentResponse[]>(
      `/releases/${releaseId}/components`
    );
    return response.data;
  }

  /**
   * Get components by release ID with effort estimates
   */
  static async getComponentsByReleaseIdWithEffortEstimates(releaseId: number): Promise<ComponentResponse[]> {
    const response = await apiClient.get<ComponentResponse[]>(
      `/releases/${releaseId}/components/with-effort-estimates`
    );
    return response.data;
  }

  /**
   * Get all components with pagination
   */
  static async getComponentsPaginated(
    page: number = 0,
    size: number = 20,
    sort?: string
  ): Promise<PaginatedResponse<ComponentResponse>> {
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString()
    });
    
    if (sort) {
      params.append('sort', sort);
    }

    const response = await apiClient.get<PaginatedResponse<ComponentResponse>>(
      `${this.BASE_URL}?${params.toString()}`
    );
    return response.data;
  }

  /**
   * Search components by name
   */
  static async searchComponentsByName(name: string): Promise<ComponentResponse[]> {
    const response = await apiClient.get<ComponentResponse[]>(
      `${this.BASE_URL}/search?name=${encodeURIComponent(name)}`
    );
    return response.data;
  }

  /**
   * Get component types (enum values)
   */
  static getComponentTypes(): { value: ComponentTypeEnum; label: string }[] {
    return [
      { value: ComponentType.ETL as ComponentTypeEnum, label: 'ETL' },
      { value: ComponentType.FORGEROCK_IGA as ComponentTypeEnum, label: 'ForgeRock IGA' },
      { value: ComponentType.FORGEROCK_UI as ComponentTypeEnum, label: 'ForgeRock UI' },
      { value: ComponentType.FORGEROCK_IG as ComponentTypeEnum, label: 'ForgeRock IG' },
      { value: ComponentType.FORGEROCK_IDM as ComponentTypeEnum, label: 'ForgeRock IDM' },
      { value: ComponentType.SAILPOINT as ComponentTypeEnum, label: 'SailPoint' },
      { value: ComponentType.FUNCTIONAL_TEST as ComponentTypeEnum, label: 'Functional Test' }
    ];
  }

  /**
   * Validate component data
   */
  static validateComponent(component: ComponentRequest): { isValid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (!component.name || component.name.trim().length === 0) {
      errors.push('Component name is required');
    } else if (component.name.length > 100) {
      errors.push('Component name must not exceed 100 characters');
    }

    if (!component.componentType) {
      errors.push('Component type is required');
    }

    if (component.technicalDesignDays < 1 || component.technicalDesignDays > 1000) {
      errors.push('Technical design days must be between 1 and 1000');
    }

    if (component.buildDays < 1 || component.buildDays > 1000) {
      errors.push('Build days must be between 1 and 1000');
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  /**
   * Calculate total effort days for a component
   */
  static calculateTotalEffortDays(component: Component | ComponentRequest): number {
    return component.technicalDesignDays + component.buildDays;
  }
}

// Export default instance for convenience
export default ComponentService;

File: relmgmt/backend/src/main/java/com/polycoder/relmgmt/service/ComponentService.java
package com.polycoder.relmgmt.service;

import com.polycoder.relmgmt.dto.ComponentRequest;
import com.polycoder.relmgmt.dto.ComponentResponse;
import com.polycoder.relmgmt.entity.ComponentTypeEnum;

import java.util.List;

/**
 * Service interface for Component business logic operations.
 */
public interface ComponentService {
    
    /**
     * Find all components for a scope item.
     * 
     * @param scopeItemId the scope item ID
     * @return list of component responses
     */
    List<ComponentResponse> findByScopeItemId(Long scopeItemId);
    
    /**
     * Find a component by ID.
     * 
     * @param id the component ID
     * @return component response
     * @throws com.polycoder.relmgmt.exception.ResourceNotFoundException if component not found
     */
    ComponentResponse findById(Long id);
    
    /**
     * Create a new component for a scope item.
     * 
     * @param scopeItemId the scope item ID
     * @param request the component request
     * @return created component response
     * @throws com.polycoder.relmgmt.exception.ResourceNotFoundException if scope item not found
     * @throws com.polycoder.relmgmt.exception.ValidationException if validation fails
     */
    ComponentResponse create(Long scopeItemId, ComponentRequest request);
    
    /**
     * Update an existing component.
     * 
     * @param id the component ID
     * @param request the component request
     * @return updated component response
     * @throws com.polycoder.relmgmt.exception.ResourceNotFoundException if component not found
     * @throws com.polycoder.relmgmt.exception.ValidationException if validation fails
     */
    ComponentResponse update(Long id, ComponentRequest request);
    
    /**
     * Delete a component.
     * 
     * @param id the component ID
     * @throws com.polycoder.relmgmt.exception.ResourceNotFoundException if component not found
     */
    void delete(Long id);
    
    /**
     * Check if a component can be deleted (no dependencies).
     * 
     * @param id the component ID
     * @return true if component can be deleted
     */
    boolean canDeleteComponent(Long id);
    
    /**
     * Find components by component type.
     * 
     * @param componentType the component type
     * @return list of component responses
     */
    List<ComponentResponse> findByComponentType(ComponentTypeEnum componentType);
    
    /**
     * Find components by scope item ID and component type.
     * 
     * @param scopeItemId the scope item ID
     * @param componentType the component type
     * @return list of component responses
     */
    List<ComponentResponse> findByScopeItemIdAndComponentType(Long scopeItemId, ComponentTypeEnum componentType);
    
    /**
     * Count components by scope item ID.
     * 
     * @param scopeItemId the scope item ID
     * @return the count
     */
    long countByScopeItemId(Long scopeItemId);
    
    /**
     * Find components by release ID.
     * 
     * @param releaseId the release ID
     * @return list of component responses
     */
    List<ComponentResponse> findByReleaseId(Long releaseId);
    
    /**
     * Find components by release ID and component type.
     * 
     * @param releaseId the release ID
     * @param componentType the component type
     * @return list of component responses
     */
    List<ComponentResponse> findByReleaseIdAndComponentType(Long releaseId, ComponentTypeEnum componentType);

    /**
     * Find components by release ID with effort estimates.
     * 
     * @param releaseId the release ID
     * @return list of component responses with effort estimates
     */
    List<ComponentResponse> findByReleaseIdWithEffortEstimates(Long releaseId);
}
